include/hnswlib/space_pq.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+#include "hnswlib.h"
+#include <stdint.h>
+
+namespace hnswlib {
+
+typedef struct pq_dist_param_s {
+  size_t n_subvectors;
+  size_t n_clusters;
+  float *dist_mat;
+} pq_dist_param_t;
+
+template <typename CODETYPE>
+static float PQLookup(const void *pVect1v, const void *pVect2v,
+                      const void *qty_ptr) {
+  CODETYPE *pVect1 = (CODETYPE *)pVect1v;
+  CODETYPE *pVect2 = (CODETYPE *)pVect2v;
+  pq_dist_param_t *qty = (pq_dist_param_t *)qty_ptr;
+  size_t subspace_mat_size = (qty->n_clusters) * (qty->n_clusters);
+  size_t n_clusters = qty->n_clusters;
+  float res = 0;
+
+  for (size_t i = 0; i < qty->n_subvectors; i++) {
+    res += qty->dist_mat[i * subspace_mat_size + (*pVect1) * n_clusters +
+                         (*pVect2)];
+    pVect1++;
+    pVect2++;
+  }
+  return (res);
+}
+
+template <typename CODETYPE> class PQ_L2Space : public SpaceInterface<float> {
+  DISTFUNC<float> fstdistfunc_;
+  size_t data_size_, d_subvectors;
+  pq_dist_param_t param;
+  float *codebook;
+
+public:
+  PQ_L2Space(size_t n_subvectors, size_t n_clusters, size_t d_subvectors,
+             float *codebook) {
+    param.n_subvectors = n_subvectors;
```

Comment:
why do we need a pointer to a float and  not a float?

Reply:
`codebook` is a 2-dimensional array with float elements.

Reply:
and who owns it?

Reply:
It comes from a buffer held by `pq_codec`, which is declared in:
https://github.com/jina-ai/annlite/blob/103b5bf31b59ca041d7ca70b8e0442421bcdf279/annlite/core/codec/pq.py#L78-L80

The collection for this buffer is supposed to be controlled by the Python interpreter I think.